server/socketManager.js

```diff
@@ -0,0 +1,333 @@
+const { Server } = require('bweb');
+const assert = require('bsert');
+const { sha256, random, ccmp } = require('bcrypto');
+const { base58 } = require('bstring');
+const { URL } = require('url');
+const Validator = require('bval');
+
+class SocketManager extends Server {
+  /**
+   * Create an http server.
+   * @constructor
+   * @param {Object} options
+   */
+  constructor(options) {
+    super(new SocketManagerOptions(options));
+
+    this.logger =
+      this.options.logger && this.options.logger.context('socket-manager');
+    this.clients = new Map();
+    this.types = ['node', 'wallet'];
+    this.init();
+  }
+
+  /*
+   * Initialize server
+   */
+  init() {
+    this.on('request', req => {
+      if (req.method === 'POST' && req.pathname === '/') return;
+
+      this.logger.debug(
+        'Request for method=%s path=%s (%s).',
+        req.method,
+        req.pathname,
+        req.socket.remoteAddress
+      );
+    });
+
+    this.on('listening', address => {
+      this.logger.info(
+        'SocketManager server listening on %s (port=%d).',
+        address.address,
+        address.port
+      );
+    });
+  }
+
+  /*
+   * handle new socket clients that connect to server
+   * @param {WebSocket} socket
+   * @returns {null}
+   */
+  handleSocket(socket) {
+    socket.hook('auth', (...args) => {
+      if (socket.channel('auth')) throw new Error('Already authed.');
+
+      if (!this.options.noAuth) {
+        const valid = new Validator(args);
+        const key = valid.str(0, '');
+        if (key.length > 255) throw new Error('Invalid API key.');
+
+        const data = Buffer.from(key, 'utf8');
+        const hash = sha256.digest(data);
+
+        if (!ccmp(hash, this.options.apiHash))
+          throw new Error('Invalid API key.');
+      }
+
+      socket.join('auth');
+      this.logger.info('Successful auth from %s.', socket.host);
+      this.handleAuth(socket);
+      return null;
+    });
+  }
+
+  /*
+   * Utility to verify that the pathname
+   * matches with the id of a client on the class
+   * @private
+   * @param {string} pathname - pathname to validate
+   * @returns {string} id
+   */
+  getIdFromPath(pathname) {
+    assert(
+      pathname[0] === '/' && pathname[pathname.length - 1] === '/',
```

Comment:
Looking at the nodejs URL docs, https://nodejs.org/api/url.html#url_url_pathname,
it doesn't show the last char being a `/`, are you certain that it will always be
parsed in that way no matter what the user initializes their client with? ie
`new WS({ path: '/foo/' })` vs `new WS({ path: '/foo' })` - I think that including the trailing
slash is not standardized between different clients but I could be wrong

Reply:
not 100% but the sockets from what I've seen add a trailing `/` and a query string `?=transport...`